govcd/user.go

```diff
@@ -0,0 +1,456 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+)
+
+// Definition of an OrgUser
+type OrgUser struct {
+	User     *types.User
+	client   *Client
+	AdminOrg *AdminOrg // needed to be able to update, as the list of roles is found in the Org
+}
+
+// Simplified structure to insert or modify an organization user
+type OrgUserConfiguration struct {
+	Name            string // Mandatory
+	Password        string // Mandatory
+	RoleName        string // Mandatory
+	ProviderType    string // Optional: defaults to "INTEGRATED"
+	IsEnabled       bool   // Optional: defaults to false
+	IsLocked        bool   // Only used for updates
+	DeployedVmQuota int    // Optional: 0 means "unlimited"
+	StoredVmQuota   int    // Optional: 0 means "unlimited"
+	FullName        string
+	Description     string
+	EmailAddress    string
+	Telephone       string
+}
+
+const (
+	// Common role names and provider types are kept here to reduce hard-coded text and prevent mistakes
+	// Roles that are added to the organization need to be entered as free text
+
+	OrgUserRoleOrganizationAdministrator = "Organization Administrator"
+	OrgUserRoleCatalogAuthor             = "Catalog Author"
+	OrgUserRoleVappAuthor                = "vApp Author"
+	OrgUserRoleVappUser                  = "vApp User"
+	OrgUserRoleConsoleAccessOnly         = "Console Access Only"
+	OrgUserRoleDeferToIdentityProvider   = "Defer to Identity Provider"
+
+	// Allowed values for provider types
+	OrgUserProviderIntegrated = "INTEGRATED" // The user is created locally or imported from LDAP
+	OrgUserProviderSAML       = "SAML"       // The user is imported from a SAML identity provider.
+	OrgUserProviderOAUTH      = "OAUTH"      // The user is imported from an OAUTH identity provider
+)
+
+// Used to check the validity of provider type on creation
+var OrgUserProviderTypes = []string{
+	OrgUserProviderIntegrated,
+	OrgUserProviderSAML,
+	OrgUserProviderOAUTH,
+}
+
+// NewUser creates an empty user
+func NewUser(cli *Client, org *AdminOrg) *OrgUser {
+	return &OrgUser{
+		User:     new(types.User),
+		client:   cli,
+		AdminOrg: org,
+	}
+}
+
+// GetUserByNameOrId retrieves an user within an admin organization
+// by either name or ID
+// Returns a valid user if it exists. If it doesn't, returns nil and ErrorEntityNotFound
+func (adminOrg *AdminOrg) GetUserByNameOrId(identifier string, willRefresh bool) (*OrgUser, error) {
```

Comment:
Seems here we explicitly state that `identifier` can hold name of ID. Or maybe have seperata method GetUserByNameOrId(identifier string, identifier name, refresh bool).. O wrappers - `GetUserByName, GetUserById`

Reply:
As I mentioned above, it's unlikely that we can mistake an ID as a name, and for OrgUsers it is impossible, as names can't contain some of the characters we find in IDs.

I don't see the risk of this approach: even if it were possible to use a name that looks like `urn:vcloud:user:440a94b2-414b-4923-aff7-439603c8dd48`, what's the risk, since we check both fields and return a match if they are the same?

Reply:
I just personally don't like implicit behaviour - double meaning parameters. Especially in public API. If everyone else ok with this - I can live with it.

Reply:
It is not implicit. It does what it says. Finds "By name or by ID".
This is also on line with what I propose in https://github.com/vmware/go-vcloud-director/issues/211

Reply:
I would like opinion @lvirbalas and @Didainius here

Reply:
IMHO probably separate fields would be clearer here.